File: WorkflowPatternFinder/Gensim/ProcessTree.py
import sys
from Node import *

class ProcessTree(object):
    """A Process Tree object."""

    def __init__(self, filePath, id, rootId):
        self._filePath = filePath
        self._id = id
        self._rootId = rootId
        self._root = None
        self._nodes = []

    def AddNode(self, node):
        self._nodes.append(node)
        if node.GetId() == self._rootId:
            node.SetRoot(True)
            self._root = node
    
    def GetNode(self, id):
        for node in self._nodes:
            if node.GetId() == id:
                return node
        raise ValueError('This node cannot be found!')
  
    def GetNodes(self):
        return self._nodes

    def GetId(self):
        return self._id
      
    def GetRoot(self):
        return self._root

    def GetTreeSize(self):
        return len(self._nodes)
    
    def SetParentalRelation(self, parentId, childId):
        parent = self.GetNode(parentId)
        child = self.GetNode(childId)
        parent.AddChild(child)
        child.SetParent(parent)

    def SetNumbers(self):
        nodelist = [self.GetRoot()]
        number = 0
        while any(nodelist):
          currentNode = nodelist.pop(0)
          currentNode.SetNumber(number)
          for child in currentNode.GetChildren():
            nodelist.append(child)
          number += 1

    def RemoveNode(self, node):
        if node.GetType() != "ManualTask":
          throw("Removing this node is not allowed!")
        if not node.IsRoot():
            parent = node.GetParent()
            parent.RemoveChild(node)
        self._nodes.remove(node)

    def GetPostOrder(self):
      data = []
      def recurse(node):
        children = node.GetChildren()
        for child in children:
          recurse(child)
        data.append(node)
      recurse(self.GetRoot())
      return data
        
      
                

File: WorkflowPatternFinder/WorkflowEventLogProcessor/WorkflowEventLogProcessor/WorkflowEventLogProcessor/ProcessTree.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace WorkflowEventLogProcessor
{
  public class ProcessTree
  {
    private readonly string _filePath;
    private readonly Guid _id;
    private readonly Guid _rootId;
    private Node _root;
    private readonly List<Node> _nodes = new List<Node>();

    public ProcessTree(string filePath, string id, string rootId)
    {
      Guid.TryParseExact(id, "D", out _id);
      Guid.TryParseExact(rootId, "D", out _rootId);
      _filePath = filePath;
    }

    public void AddNode(Node node)
    {
      _nodes.Add(node);
      if(node.GetId() == _rootId)
      {
        node.SetRoot(true);
        _root = node;
      }
    }

    private Node GetNode(Guid id)
    {
      try
      {
        return _nodes.Single(n => n.GetId() == id);
      }
      catch(InvalidOperationException e)
      {
        throw new InvalidOperationException($"The Guid {id} you look for is not unique!", e);
      }
    }

    public Guid GetId()
    {
      return _id;
    }

    public Node GetRoot()
    {
      return _root;
    }

    public void SetParentalRelation(Guid parentId, Guid childId)
    {
      Node parent = GetNode(parentId);
      Node child = GetNode(childId);
      parent.AddChild(child);
      child.SetParent(parent);
    }

    public string GetFilePath()
    {
      return _filePath;
    }
  }
}